### PR TITLE
Externalize PSP admission controller

### PR DESCRIPTION
**Pull request by yue9944882**
continue with previous work by @xmudrii #67846. 
sadly he's recently busy to keep track of this thread.

/kind cleanup
/sig api-machinery

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/security/podsecuritypolicy/util/util.go [hunk 1]
```diff
@@ -22,7 +22,8 @@ import (
 
 	"k8s.io/apimachinery/pkg/util/sets"
 	api "k8s.io/kubernetes/pkg/apis/core"
-	"k8s.io/kubernetes/pkg/apis/policy"
+	//"k8s.io/kubernetes/pkg/apis/policy"
```
Comment (caesarxuchao) [inline pkg/security/podsecuritypolicy/util/util.go]: remove it?

File: pkg/security/podsecuritypolicy/util/util.go [hunk 2]
```diff
@@ -40,102 +41,102 @@ func GetAllFSTypesExcept(exceptions ...string) sets.String {
...
+		string(policyv1beta1.ISCSI),
+		string(policyv1beta1.Glusterfs),
+		string(policyv1beta1.PersistentVolumeClaim),
+		string(policyv1beta1.RBD),
+		string(policyv1beta1.Cinder),
+		string(policyv1beta1.CephFS),
+		string(policyv1beta1.DownwardAPI),
+		string(policyv1beta1.FC),
+		string(policyv1beta1.ConfigMap),
+		//string(policyv1beta1.VsphereVolume),
```
Comment (caesarxuchao) [inline pkg/security/podsecuritypolicy/util/util.go]: I understand that you commented it out because the const didn't exist in v1beta1. I think we have to add them to v1beta1 first, then merge this PR.

File: pkg/security/podsecuritypolicy/util/util.go (hunk 1 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/util/util.go]: done

File: pkg/security/podsecuritypolicy/util/util.go (hunk 2 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/util/util.go]: its done by #69704 #69795

**Comment by liggitt** [comment]
please leave the security/podsecuritypolicy/OWNERS file at the parent location

File: pkg/security/podsecuritypolicy/v1beta1/factory.go [hunk 3]
```diff
@@ -22,16 +22,16 @@ import (
...
-	api "k8s.io/kubernetes/pkg/apis/core"
-	"k8s.io/kubernetes/pkg/apis/policy"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/apparmor"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/capabilities"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/group"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/seccomp"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/selinux"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/sysctl"
-	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/user"
+	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/v1beta1/apparmor"
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/v1beta1/factory.go]: unless we intend to have multiple versions of the podsecuritypolicy utilities (which I don't think we do), let's leave these in the existing location. as PSP graduates to v1, we would wait N releases, then switch the admission plugin to use v1

File: pkg/security/podsecuritypolicy/v1beta1/factory.go [hunk 4]
```diff
@@ -110,88 +110,88 @@ func (f *simpleStrategyFactory) CreateStrategies(psp *policy.PodSecurityPolicy,
 }
 
 // createUserStrategy creates a new user strategy.
-func createUserStrategy(opts *policy.RunAsUserStrategyOptions) (user.RunAsUserStrategy, error) {
+func createUserStrategy(opts *policyv1beta1.RunAsUserStrategyOptions) (user.RunAsUserStrategy, error) {
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/v1beta1/factory.go]: I'd prefer leaving the import named `policy`, and pointing it to `policy/v1beta1`, so that graduating this code to point to `policy/v1` would be as simple as changing the import (assuming no other changes between v1beta1 and v1).

Comment (liggitt) [reply pkg/security/podsecuritypolicy/v1beta1/factory.go]: that comment applies throughout. referring to core/v1 as 'corev1' is fine, since that package is already at GA

File: pkg/security/podsecuritypolicy/v1beta1/provider.go [hunk 5]
```diff
@@ -339,19 +339,15 @@ func (s *simpleProvider) ValidateContainer(pod *api.Pod, container *api.Containe
 	}
 
 	allowEscalation := sc.AllowPrivilegeEscalation()
-	if !s.psp.Spec.AllowPrivilegeEscalation && allowEscalation == nil {
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/v1beta1/provider.go]: let's not mix changes like this into this PR. defaulting ensures AllowPrivilegeEscalation is non-nil. if you want to pursue this anyway, let's open a separate PR to discuss it.

File: pkg/security/podsecuritypolicy/v1beta1/provider.go [hunk 6]
```diff
@@ -339,19 +339,15 @@ func (s *simpleProvider) ValidateContainer(pod *api.Pod, container *api.Containe
...
+	if (s.psp.Spec.AllowPrivilegeEscalation != nil && !*s.psp.Spec.AllowPrivilegeEscalation) && (allowEscalation == nil || *allowEscalation) {
 		allErrs = append(allErrs, field.Invalid(scPath.Child("allowPrivilegeEscalation"), allowEscalation, "Allowing privilege escalation for containers is not allowed"))
 	}
 
-	if !s.psp.Spec.AllowPrivilegeEscalation && allowEscalation != nil && *allowEscalation {
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/v1beta1/provider.go]: same here, pull this change out to a separate PR

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go [hunk 7]
```diff
@@ -256,15 +258,19 @@ func (c *PodSecurityPolicyPlugin) computeSecurityContext(a admission.Attributes,
 	)
 
 	for _, provider := range providers {
-		podCopy := pod.DeepCopy()
+		externalPod := &corev1.Pod{}
+		if err := k8s_api_v1.Convert_core_Pod_To_v1_Pod(pod, externalPod, nil); err != nil {
```
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: using the external PSP objects makes sense (because the admission plugin is acting like an API client, watching those policy objects via an informer), but why are we changing the admission interfaces and utilities to work with external v1 pod/container objects, and then doing conversions here? I expected the pod/container references to remain internal. 

cc @deads2k

File: pkg/security/podsecuritypolicy/v1beta1/factory.go (hunk 3 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/v1beta1/factory.go]: make sense. also, keeping the same location makes the pull much smaller.

File: pkg/security/podsecuritypolicy/v1beta1/factory.go (hunk 4 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/v1beta1/factory.go]: done

File: pkg/security/podsecuritypolicy/v1beta1/provider.go (hunk 5 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/v1beta1/provider.go]: > defaulting ensures AllowPrivilegeEscalation is non-nil. 

i added this to prevent nil panic. but yeah, ci still looks happy w/o this.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go (hunk 7 above)
Comment (yue9944882) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: how about now? ptal

File: pkg/security/podsecuritypolicy/selinux/mustrunas_test.go [hunk 8]
```diff
@@ -54,13 +56,16 @@ func TestMustRunAsOptions(t *testing.T) {
...
 
 func TestMustRunAsGenerate(t *testing.T) {
+	externalSELinuxOptions := &corev1.SELinuxOptions{
+		User:  "user",
+		Role:  "role",
+		Type:  "type",
+		Level: "level",
+	}
+	internalSELinuxOptions := &api.SELinuxOptions{}
+	v1.Convert_v1_SELinuxOptions_To_core_SELinuxOptions(externalSELinuxOptions, internalSELinuxOptions, nil)
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/selinux/mustrunas_test.go]: I didn't expect conversions here, just a change to working with external objects

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go (hunk 7 above)
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: externalPod is unused now, right? can we drop this change?

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go [hunk 9]
```diff
@@ -276,7 +282,7 @@ func (c *PodSecurityPolicyPlugin) computeSecurityContext(a admission.Attributes,
 		switch {
 		case !mutated:
 			// if it validated without mutating anything, use this result
-			return podCopy, provider.GetPSPName(), nil, nil
+			return pod, provider.GetPSPName(), nil, nil
```
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: revert this change

File: pkg/security/podsecuritypolicy/capabilities/capabilities_test.go [hunk 10]
```diff
@@ -279,7 +278,7 @@ func TestValidateAdds(t *testing.T) {
 			},
 		},
 		"no required, all allowed, container requests valid": {
-			allowedCaps: []api.Capability{policy.AllowAllCapabilities},
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/capabilities/capabilities_test.go]: updated

File: pkg/security/podsecuritypolicy/selinux/mustrunas.go [hunk 11]
```diff
@@ -21,14 +21,14 @@ import (
...
 
+	policy "k8s.io/api/policy/v1beta1"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	api "k8s.io/kubernetes/pkg/apis/core"
-	"k8s.io/kubernetes/pkg/apis/policy"
 	"k8s.io/kubernetes/pkg/security/podsecuritypolicy/util"
 )
 
 type mustRunAs struct {
-	opts *policy.SELinuxStrategyOptions
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/selinux/mustrunas.go]: it's involving unnecessary dependency. if we keep this, this would be the last place referencing internal psp package.  for a complete switching, i think we should slightly refactor here

File: pkg/security/podsecuritypolicy/selinux/mustrunas_test.go [hunk 12]
```diff
@@ -154,7 +161,9 @@ func TestMustRunAsValidate(t *testing.T) {
 			continue
 		}
 
-		errs := mustRunAs.Validate(nil, nil, nil, tc.podSeLinux)
+		internalSELinuxOptions := api.SELinuxOptions{}
+		v1.Convert_v1_SELinuxOptions_To_core_SELinuxOptions(tc.podSeLinux, &internalSELinuxOptions, nil)
```
Comment (yue9944882) [inline pkg/security/podsecuritypolicy/selinux/mustrunas_test.go]: @liggitt as you can see, throughout this psp externalization, we are putting all the conversion logic into the constructor function of each strategy. so now, we list/watch external psp objects in admission controller, and do internal conversion when instantiating strategies (so that we neednt do conversion everytime we actually call strategy), finally the psp provider interface will accept internal pod/container and return internal types back to admission controllers. this will be how externalized psp works. wdyt?

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go (hunk 9 above)
Comment (yue9944882) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: done

File: pkg/security/podsecuritypolicy/selinux/mustrunas_test.go (hunk 8 above)
Comment (yue9944882) [reply pkg/security/podsecuritypolicy/selinux/mustrunas_test.go]: how about now? the conversion is replaced by struct instantiation

File: pkg/security/podsecuritypolicy/selinux/mustrunas_test.go (hunk 12 above)
Comment (liggitt) [reply pkg/security/podsecuritypolicy/selinux/mustrunas_test.go]: > do internal conversion when instantiating strategies (so that we neednt do conversion everytime we actually call strategy)

strategies are instantiated every time a pod is admitted. doing conversions as part of their construction is not what we want.

File: pkg/security/podsecuritypolicy/selinux/mustrunas.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/security/podsecuritypolicy/selinux/mustrunas.go]: switching from internal to external is fine, the change from SELinuxStrategyOptions to SELinuxOptions is what surprised me. I see now the mismatch in what we get from the PSP object and what the Generate() function returns (and the comparisons made in Validate()). I guess the conversion here in construction is ok

File: pkg/security/podsecuritypolicy/selinux/mustrunas_test.go [hunk 13]
```diff
@@ -70,22 +72,27 @@ func TestMustRunAsGenerate(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unexpected error generating selinux %v", err)
 	}
-	if !reflect.DeepEqual(generated, opts.SELinuxOptions) {
+	if !reflect.DeepEqual(generated, &api.SELinuxOptions{
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/selinux/mustrunas_test.go]: reconstructing like this means we're vulnerable to drift if new fields are added. prefer converting to make sure we catch everything

File: pkg/security/podsecuritypolicy/capabilities/capabilities.go [hunk 14]
```diff
@@ -36,11 +37,23 @@ var _ Strategy = &defaultCapabilities{}
 
 // NewDefaultCapabilities creates a new defaultCapabilities strategy that will provide defaults and validation
 // based on the configured initial caps and allowed caps.
-func NewDefaultCapabilities(defaultAddCapabilities, requiredDropCapabilities, allowedCaps []api.Capability) (Strategy, error) {
+func NewDefaultCapabilities(defaultAddCapabilities, requiredDropCapabilities, allowedCaps []corev1.Capability) (Strategy, error) {
+	internalDefaultAddCaps := []api.Capability{}
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/capabilities/capabilities.go]: initialize these to the length of the slice you're copying in to avoid reallocation while appending

**Comment by liggitt** [comment]
one comment on initializing slice length during copy, squash commits, then LGTM

File: pkg/security/podsecuritypolicy/capabilities/capabilities.go [hunk 15]
```diff
@@ -36,11 +37,23 @@ var _ Strategy = &defaultCapabilities{}
 
 // NewDefaultCapabilities creates a new defaultCapabilities strategy that will provide defaults and validation
 // based on the configured initial caps and allowed caps.
-func NewDefaultCapabilities(defaultAddCapabilities, requiredDropCapabilities, allowedCaps []api.Capability) (Strategy, error) {
+func NewDefaultCapabilities(defaultAddCapabilities, requiredDropCapabilities, allowedCaps []corev1.Capability) (Strategy, error) {
+	internalDefaultAddCaps := make([]api.Capability, len(defaultAddCapabilities))
+	for i := 0; i < len(defaultAddCapabilities); i++ {
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/capabilities/capabilities.go]: this form would be easier to read:

```suggestion
	for i, c := range defaultAddCapabilities {
		internalDefaultAddCaps[i] = api.Capability(c)
	}
```
